Update dev containers docs

bamurtaugh opened the pull request:
Adding documentation based on https://github.com/microsoft/vscode-remote-release/issues/5442.

- Document new dev container features
- Document how pre-building generally works, and how it relates to features and the `devcontainer CLI`

Would love to get your thoughts on content, formatting, anything else we should include, etc. Thanks!

cc @Chuxel @chrmarti @2percentsilk

Chuxel commented on docs/remote/containers.md:
```diff
@@ -98,7 +98,11 @@ This quick start covers how to set up a dev container for an existing project to
-    The list will be automatically sorted based on the contents of the folder you open. Note the dev container definitions displayed come from the [vscode-dev-containers repository](https://aka.ms/vscode-dev-containers). You can browse the `containers` folder of that repository to see the contents of each definition.
+    The list will be automatically sorted based on the contents of the folder you open.
+
+    You will also be able to customize your dev container with additional features, which [you can read more about below](#Dev-container-features).
```
We should probably say you "may" be able to pick from features since not all of the definitions have them.

Chuxel commented on docs/remote/containers.md:
````diff
@@ -231,6 +235,36 @@ Selecting the **Remote-Containers: Add Development Container Configuration Files
+```
+
+The features are sourced from the [script library](https://github.com/microsoft/vscode-dev-containers/tree/main/script-library) in the vscode-dev-containers repo.
+
````
We could also highlight that there is a "Configure Container Features" command to modify the list after you've added a definition to your folder.

chrmarti replied:
Maybe mention that these are also available with IntelliSense in the file itself.

bamurtaugh replied:
Both great points - added info about the configure command and IntelliSense.

Chuxel commented on docs/remote/containers.md:
````diff
@@ -231,6 +235,36 @@ Selecting the **Remote-Containers: Add Development Container Configuration Files
+	}
+```
+
+The features are sourced from the [script library](https://github.com/microsoft/vscode-dev-containers/tree/main/script-library) in the vscode-dev-containers repo.
````
Maybe we should point to the /docs folder for now and add a "usage" step to those docs to use it as a feature?  What do you think @2percentsilk?

bamurtaugh replied:
I think that makes sense - updated to point to the /docs folder.

Chuxel commented on docs/remote/containers.md:
```diff
@@ -231,6 +235,36 @@ Selecting the **Remote-Containers: Add Development Container Configuration Files
+
+The features are sourced from the [script library](https://github.com/microsoft/vscode-dev-containers/tree/main/script-library) in the vscode-dev-containers repo.
+
+## Pre-building
```
Since "Pre-building" is unfortunatley getting used in a slightly different way in Codespaces, I'd suggest we be specific and say "Pre-building dev container images"

Chuxel commented on docs/remote/containers.md:
```diff
@@ -231,6 +235,36 @@ Selecting the **Remote-Containers: Add Development Container Configuration Files
+* Create a dev container configuration, customizing as you wish (including [features](#Dev-container-features))
+* Use the `devcontainer CLI` to build your image (the `devcontainer CLI` supports building images with features)
+* Push your image: `docker push <your_image_name>`
+
```
I'd suggest a code snippet illustrating the commands and mentioning that you can push your image to a container registry like the Azure Continer Registry or GitHub Container Registry (with links to docs for each).

```
$ devcontainer build --image-name your-registry.azurecr.io/your-image-name
$ docker push your-registry.azurecr.io/your-image-name
```

We also need to tell people how to install the dev dcontainer CLI w/o VS Code now that this is possible (though that could happen over on the CLI page.)

chrmarti replied:
I have added basic installation instructions in the CLI's README.md. These will show up on the package's page next time we publish the package. https://github.com/microsoft/vscode-remote-containers/blob/main/dev-containers-user-cli/README.md

bamurtaugh replied:
Updated this section with example commands like Chuck lists, more links to docs, and info from the CLI's readme like Christof mentioned. Let me know what you think!

Chuxel replied:
@bamurtaugh So based on today's discussion, we'll need to update the instructions for installing the CLI, but generally something along these lines makes sense.

I'd suggest we move the general install instructions for the CLI into the existing dev container CLI document we have that describes the command you can use.  We could have "install using VS Code" and "Install from the command line" as two install options

Chuxel commented on docs/remote/containers.md:
```diff
@@ -231,6 +235,36 @@ Selecting the **Remote-Containers: Add Development Container Configuration Files
+* Create a source code repository
+* Create a dev container configuration, customizing as you wish (including [features](#Dev-container-features))
+* Use the `devcontainer CLI` to build your image (the `devcontainer CLI` supports building images with features)
+* Push your image: `docker push <your_image_name>`
```
We should add in that you can then change the dev container in your source repository to reference the image directly with perhaps an example `image` reference in devcontainer.json and docker compose.

bamurtaugh replied:
Added a sentence around this. I didn't add an example dev container as it feels that might crowd the section and I think the statement itself makes sense? But if you think a greater example would help, or any additional info, just let me know.

Chuxel commented:
Made a few suggestions.

Might we also want to mention features in the article on creating a dev container?

bamurtaugh commented:
> Made a few suggestions.

@chuxel Thanks so much! I don't seem to see them on my end - are they in a review?

> Might we also want to mention features in the article on creating a dev container?

Great call, I can update that section.

Chuxel commented:
@bamurtaugh Yeah they were.  I was using github.dev and did  an ad-hoc comment instead of submitting my review. 🤦‍♂️

bamurtaugh commented:
@Chuxel @chrmarti Thank you both for your great feedback! Did a round of updates - if you have a moment, would love to get your thoughts.

Chuxel commented:
@bamurtaugh Made one comment on the updates for the installation steeps for the CLI above, but the other thing of note is that it sounds like Codespaces won't get the CLI upddates in place for Wednesday.  There's a 

@2percentsilk @joshspicer and I were thinking that we may want to indicate both here and in the release notes that features support for Codespaces is "coming soon" so that anyone trying when the release notes first go out aren't confused / disappointed.

bamurtaugh commented on docs/remote/devcontainer-cli.md:
````diff
@@ -37,6 +39,21 @@ To install the devcontainer CLI:
       -h, --help  Show help  [boolean]
     ```
 
+### Install devcontainer CLI outside VS Code
````
@Chuxel Is this the area you're referring to in https://github.com/microsoft/vscode-docs/pull/4856#discussion_r722510301? i.e. we can update this section to be called "Install from the command line" and include the latest instructions on how to install the CLI?

But otherwise the content in the pre-building section makes sense?

Chuxel replied:
@bamurtaugh Yep! Christof also illustrated how to use it via npx.  So there's two ways to install and use the CLI:

Global install:
```
$ npm install -g https://github.com/microsoft/vscode-remote-release/releases/latest/download/dev-container-cli.tgz
$ devcontainer --help
```

Local install:
```
$ npm install https://github.com/microsoft/vscode-remote-release/releases/latest/download/dev-container-cli.tgz
$ npx @microsoft/dev-container-cli --help
```

Using npx causes it to use the package name.

To future proof a bit, we could aka.ms that URL and reference that instead.

bamurtaugh replied:
Thanks so much for this! Created https://aka.ms/dev-container-cli and updated this section.

bamurtaugh commented:
@Chuxel I think I updated the current outstanding feedback! 

- Updated the install section in devcontainer-cli.md (titles of the sections, command line instructions for global and local, use aka.ms link)
- Added notes to containers.md and create-dev-container.md about features support coming soon to Codespaces

Chuxel commented on docs/remote/devcontainerjson-reference.md:
```diff
@@ -63,6 +63,7 @@ You may need different commands to be run at different points in the container's
 | `onCreateCommand` | string,<br>array | A command to run when creating the container. If this is a single string, it will be run in a shell. If this is an array of strings, it will be run as a single command without shell. |
 | `userEnvProbe` | enum | Indicates the type of shell VS Code should use to "probe" for user environment variables to use by default while debugging or running a task: `none` (default), `interactiveShell`, `loginShell`, or `loginInteractiveShell`. Interactive shells will typically include variables set in `/etc/bash.bashrc` and `.bashrc` while login shells usually include variables from these "rc ...
 | `devPort` | integer | Allows you to force a specific port that the VS Code Server should use in the container. Defaults to a random, available port. |
+| `features` | object | Scripts to install in the dev container when the container is built. |
```
Might be worth calling this out in devcontainer.json docs as well since we know we've got deep links here.

```suggestion
| `features` | object | An object of [dev container features](docs/remote/containers.md#dev-container-features) and options to be that should be added into your primary container. e.g. `"features": { "github-cli": "latest" }` <br /><br /> ⚠️ Support for features in GitHub Codespaces is coming soon. |
```

Chuxel commented:
@bamurtaugh Looks good!  One suggested add to devcontainer.json docs on this with a cross reference: https://github.com/microsoft/vscode-docs/pull/4856/files#r722668513

bamurtaugh commented:
@Chuxel Good catch! Just committed the suggested change. 🎉

2percentsilk commented on docs/remote/containers.md:
```diff
@@ -231,6 +235,51 @@ Selecting the **Remote-Containers: Add Development Container Configuration Files
+
+## Pre-building dev container images
+
+You can pre-build images with the tools you need rather than newly creating and building images for each project. Using pre-built images can be simpler and allows you to pin to a specific version of tools, avoiding potential breaks. Pre-building is especially valuable in CI processes.
```
I think this sentence works but was thinking about phrasing it along the lines of "rather than creating and building the image each time you create a container for a project." I think we can drive home a little stronger that you're saving time whenever you use this definition!

bamurtaugh replied:
I like that suggestion! Updated to what you have in quotes.

2percentsilk commented on docs/remote/containers.md:
````diff
@@ -231,6 +235,51 @@ Selecting the **Remote-Containers: Add Development Container Configuration Files
+$ docker push your-registry.azurecr.io/your-image-name
+```
+
+You can push your image to a container registry, like the [Azure Container Registry](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-docker-cli?tabs=azure-cli) or [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#pushing-container-images).
````
Maybe we just call out Docker again here? I know we put that in the push step but if I'm skimming through the steps, I may not click on the Push link and so miss that we highlighted docker there. I think it makes sense to just list Docker in the list of registries here again.

bamurtaugh replied:
Makes sense, added mention of Docker Hub.

2percentsilk commented on docs/remote/devcontainerjson-reference.md:
```diff
@@ -63,6 +63,7 @@ You may need different commands to be run at different points in the container's
 | `onCreateCommand` | string,<br>array | A command to run when creating the container. If this is a single string, it will be run in a shell. If this is an array of strings, it will be run as a single command without shell. |
 | `userEnvProbe` | enum | Indicates the type of shell VS Code should use to "probe" for user environment variables to use by default while debugging or running a task: `none` (default), `interactiveShell`, `loginShell`, or `loginInteractiveShell`. Interactive shells will typically include variables set in `/etc/bash.bashrc` and `.bashrc` while login shells usually include variables from these "rc ...
 | `devPort` | integer | Allows you to force a specific port that the VS Code Server should use in the container. Defaults to a random, available port. |
+| `features` | object | An object of [dev container features](docs/remote/containers.md#dev-container-features) and options to be added into your primary container. e.g. `"features": { "github-cli": "latest" }` <br /><br /> ⚠️ Support for features in GitHub Codespaces is coming soon. |
```
We may also want to label this as (Preview) or (Beta) here.

2percentsilk commented:
@bamurtaugh @Chuxel made a few comments - I think the biggest one is on whether we want to label this feature as Preview/Beta? I know I talked about that with Chuck so that users have an appropriate expectation that this feature may shift and is still in progress!

bamurtaugh commented:
Thanks for the comments @2percentsilk! Think I addressed it all 🎉 

I added notes to the devcontainer-cli.md section and the devcontainerjson-reference.md entry on features to mention it's in preview - I think that makes sense, but let me know if there are other thoughts.

bamurtaugh commented:
Thanks @2percentsilk!

@gregvanl I think this is good to merge / go live with the next release! (unless Chuck or Christof have any other feedback, but believe it's all been addressed)

Chuxel commented:
Yep LGTM!!!
